src/units/abilities.cpp

```diff
@@ -1537,29 +1537,30 @@ bool attack_type::special_active_impl(const_attack_ptr self_attack, const_attack
 	temporary_facing self_facing(self, self_loc.get_relative_dir(other_loc));
 	temporary_facing other_facing(other, other_loc.get_relative_dir(self_loc));
 
-	// Filter poison, plague, drain, first strike
-	if (tag_name == "drains" && other && other->get_state("undrainable")) {
-		return false;
-	}
-	if (tag_name == "plague" && other &&
-		(other->get_state("unplagueable") ||
-		 resources::gameboard->map().is_village(other_loc))) {
+	// Filter poison, plague, drain, slow, petrifies
+	// True if "whom" corresponds to "self", false if "whom" is "other"
+	bool whom_is_self = ((whom == AFFECT_SELF) || ((whom == AFFECT_EITHER) && special_affects_self(special, is_attacker)));
+	unit_const_ptr whom_is_unit = whom_is_self ? other : self;
+	map_location whom_is_loc = whom_is_self ? other_loc : self_loc;
```

Comment:
Minor, but I'd suggest `whom_unit` and `whom_loc`. The `is` was for "the unit which is `whom` **is** the same unit that's `self`".

Reply:
@stevecotton done, and i added your laststrike test.